// @digitaldefiance/reed-solomon-erasure.wasm/dist/browser.js
"use strict";
/**
 * Reed-Solomon Erasure Coding WASM Library - Browser Build
 *
 * This is a browser-only build with NO Node.js dependencies.
 * For Node.js, use the main entry point instead.
 *
 * @packageDocumentation
 * @module @digitaldefiance/reed-solomon-erasure.wasm/browser
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReedSolomonErasure = void 0;
// Cached WASM instance for singleton pattern
let cachedInstance = null;
let loadingPromise = null;
/**
 * Reed-Solomon Erasure Coding implementation using WebAssembly.
 * Browser-only build - no Node.js dependencies.
 */
class ReedSolomonErasure {
    constructor(exports) {
        this.exports = exports;
        this.memoryCache = null;
    }
    /**
     * Get or create a singleton instance.
     * Useful for applications that only need one encoder.
     *
     * @param wasmUrl - URL to the WASM file (defaults to same directory)
     */
    static async getInstance(wasmUrl) {
        if (cachedInstance) {
            return cachedInstance;
        }
        if (loadingPromise) {
            return loadingPromise;
        }
        loadingPromise = ReedSolomonErasure.fromUrl(wasmUrl || './reed_solomon_erasure_bg.wasm');
        cachedInstance = await loadingPromise;
        loadingPromise = null;
        return cachedInstance;
    }
    /**
     * Clear the singleton instance (useful for testing or memory management).
     */
    static clearInstance() {
        cachedInstance = null;
        loadingPromise = null;
    }
    /**
     * Load WASM from a URL using fetch.
     *
     * @param url - URL to the WASM file
     */
    static async fromUrl(url) {
        return ReedSolomonErasure.fromResponse(fetch(url));
    }
    /**
     * Load WASM from a fetch Response.
     * Supports streaming instantiation for better performance.
     *
     * @param source - Response or Promise<Response> from fetch
     */
    static async fromResponse(source) {
        const response = await source;
        // Try streaming instantiation first (more efficient)
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            try {
                // Clone response in case we need to fall back
                const { instance } = await WebAssembly.instantiateStreaming(response.clone());
                return new ReedSolomonErasure(instance.exports);
            }
            catch {
                // Fall through to ArrayBuffer method
            }
        }
        // Fallback: load entire ArrayBuffer
        const bytes = await response.arrayBuffer();
        return ReedSolomonErasure.fromBytes(bytes);
    }
    /**
     * Load WASM from raw bytes (ArrayBuffer or Uint8Array).
     *
     * @param bytes - The WASM binary
     */
    static fromBytes(bytes) {
        const module = new WebAssembly.Module(bytes);
        const instance = new WebAssembly.Instance(module);
        return new ReedSolomonErasure(instance.exports);
    }
    /**
     * Load WASM from a base64-encoded string.
     * Useful for embedding WASM directly in JavaScript.
     *
     * @param base64 - Base64-encoded WASM binary
     */
    static fromBase64(base64) {
        const binaryString = atob(base64);
        const bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return ReedSolomonErasure.fromBytes(bytes);
    }
    /**
     * Encode data shards to generate parity shards.
     *
     * @param shards - Contiguous buffer containing all shards (data + space for parity)
     * @param dataShards - Number of data shards
     * @param parityShards - Number of parity shards to generate
     * @returns Result code (RESULT_OK on success)
     */
    encode(shards, dataShards, parityShards) {
        const exports = this.exports;
        const shardsLength = shards.length;
        const shardsPointer = exports.__wbindgen_malloc(shardsLength);
        this.getUint8Memory().set(shards, shardsPointer);
        const shardSize = shardsLength / (dataShards + parityShards);
        const result = exports.encode(shardsPointer, shardsLength, dataShards, parityShards);
        if (result === ReedSolomonErasure.RESULT_OK) {
            shards.set(this.getUint8Memory().subarray(shardsPointer + shardSize * dataShards, shardsPointer + shardsLength), shardSize * dataShards);
        }
        exports.__wbindgen_free(shardsPointer, shardsLength);
        return result;
    }
    /**
     * Reconstruct missing data shards from available shards.
     *
     * @param shards - Contiguous buffer containing all shards
     * @param dataShards - Number of data shards
     * @param parityShards - Number of parity shards
     * @param shardsAvailable - Boolean array indicating which shards are valid
     * @returns Result code (RESULT_OK on success)
     */
    reconstruct(shards, dataShards, parityShards, shardsAvailable) {
        const exports = this.exports;
        const shardsLength = shards.length;
        const shardsPointer = exports.__wbindgen_malloc(shardsLength);
        this.getUint8Memory().set(shards, shardsPointer);
        const shardsAvailableLength = shardsAvailable.length;
        const shardsAvailablePointer = exports.__wbindgen_malloc(shardsAvailableLength);
        this.getUint8Memory().set(shardsAvailable.map((value) => value ? 1 : 0), shardsAvailablePointer);
        const shardSize = shardsLength / (dataShards + parityShards);
        const result = exports.reconstruct(shardsPointer, shardsLength, dataShards, parityShards, shardsAvailablePointer, shardsAvailableLength);
        if (result === ReedSolomonErasure.RESULT_OK) {
            shards.set(this.getUint8Memory().subarray(shardsPointer, shardsPointer + shardSize * dataShards));
        }
        exports.__wbindgen_free(shardsPointer, shardsLength);
        exports.__wbindgen_free(shardsAvailablePointer, shardsAvailableLength);
        return result;
    }
    /**
     * Get result code description.
     */
    static getResultMessage(code) {
        const messages = {
            0: 'OK',
            1: 'Too few shards',
            2: 'Too many shards',
            3: 'Too few data shards',
            4: 'Too many data shards',
            5: 'Too few parity shards',
            6: 'Too many parity shards',
            7: 'Too few buffer shards',
            8: 'Too many buffer shards',
            9: 'Incorrect shard size',
            10: 'Too few shards present for reconstruction',
            11: 'Empty shard',
            12: 'Invalid shard flags',
            13: 'Invalid index',
        };
        return messages[code] || `Unknown error code: ${code}`;
    }
    getUint8Memory() {
        let cache = this.memoryCache;
        if (cache === null || cache.buffer !== this.exports.memory.buffer) {
            cache = new Uint8Array(this.exports.memory.buffer);
            this.memoryCache = cache;
        }
        return cache;
    }
}
exports.ReedSolomonErasure = ReedSolomonErasure;
// Result codes
ReedSolomonErasure.RESULT_OK = 0;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_SHARDS = 1;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_SHARDS = 2;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_DATA_SHARDS = 3;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_DATA_SHARDS = 4;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_PARITY_SHARDS = 5;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_PARITY_SHARDS = 6;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_BUFFER_SHARDS = 7;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_BUFFER_SHARDS = 8;
ReedSolomonErasure.RESULT_ERROR_INCORRECT_SHARD_SIZE = 9;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_SHARDS_PRESENT = 10;
ReedSolomonErasure.RESULT_ERROR_EMPTY_SHARD = 11;
ReedSolomonErasure.RESULT_ERROR_INVALID_SHARD_FLAGS = 12;
ReedSolomonErasure.RESULT_ERROR_INVALID_INDEX = 13;
// Re-export for convenience
exports.default = ReedSolomonErasure;
