// @digitaldefiance/reed-solomon-erasure.wasm/dist/index.js
"use strict";
/**
 * Reed-Solomon Erasure Coding WASM Library
 *
 * Browser and Node.js compatible implementation.
 *
 * @packageDocumentation
 * @module @digitaldefiance/reed-solomon-erasure.wasm
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReedSolomonErasure = void 0;
// Detect browser environment - check for currentScript at module load time
const currentScript = typeof globalThis !== 'undefined' &&
    globalThis.document &&
    globalThis.document.currentScript;
// Check if we're in a browser environment
const isBrowser = typeof window !== 'undefined' || typeof self !== 'undefined';
// Cached singleton instance
let cachedInstance = null;
let loadingPromise = null;
/**
 * Load WASM bytes in Node.js environment using dynamic import
 * This avoids the top-level fs import that breaks browser bundlers
 */
async function loadWasmBytesNode(wasmPath) {
    // Dynamic import to avoid bundler issues
    const fs = await Promise.resolve().then(() => __importStar(require('fs')));
    return fs.readFileSync(wasmPath);
}
/**
 * Synchronously load WASM bytes in Node.js environment
 * Only call this in Node.js - will throw in browser
 */
function loadWasmBytesSyncNode(wasmPath) {
    // Use require for synchronous loading in Node.js
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const fs = require('fs');
    return fs.readFileSync(wasmPath);
}
class ReedSolomonErasure {
    constructor(exports) {
        this.exports = exports;
        this.memoryCache = null;
    }
    /**
     * Get or create a singleton instance.
     * Useful for applications that only need one encoder.
     *
     * @param wasmUrl - URL to the WASM file (browser) or path (Node.js)
     */
    static async getInstance(wasmUrl) {
        if (cachedInstance) {
            return cachedInstance;
        }
        if (loadingPromise) {
            return loadingPromise;
        }
        if (wasmUrl) {
            if (isBrowser) {
                loadingPromise = ReedSolomonErasure.fromResponse(fetch(wasmUrl));
            }
            else {
                loadingPromise = loadWasmBytesNode(wasmUrl).then(bytes => ReedSolomonErasure.fromBytes(bytes));
            }
        }
        else {
            loadingPromise = ReedSolomonErasure.fromCurrentDirectory();
        }
        cachedInstance = await loadingPromise;
        loadingPromise = null;
        return cachedInstance;
    }
    /**
     * Clear the singleton instance (useful for testing or memory management).
     */
    static clearInstance() {
        cachedInstance = null;
        loadingPromise = null;
    }
    /**
     * Automagical method that will try to detect environment (Node.js or browser)
     * and load *.wasm file from current directory.
     *
     * In browser: Uses fetch() to load the WASM file relative to the current script
     * In Node.js: Uses fs.readFileSync() to load from __dirname
     */
    static async fromCurrentDirectory() {
        if (isBrowser && currentScript) {
            // Browser with script tag
            const pathToCurrentScript = currentScript.src.split('/').slice(0, -1).join('/');
            return ReedSolomonErasure.fromResponse(fetch(`${pathToCurrentScript}/reed_solomon_erasure_bg.wasm`));
        }
        else if (isBrowser) {
            // Browser without currentScript (e.g., bundled) - try relative path
            return ReedSolomonErasure.fromResponse(fetch('./reed_solomon_erasure_bg.wasm'));
        }
        else {
            // Node.js environment - use dynamic import to avoid bundler issues
            const wasmBytes = await loadWasmBytesNode(`${__dirname}/reed_solomon_erasure_bg.wasm`);
            return ReedSolomonErasure.fromBytes(wasmBytes);
        }
    }
    /**
     * For asynchronous instantiation, primarily in Browser environment.
     * Expects you to load WASM file with `fetch()`.
     *
     * @param source - A Response or Promise<Response> from fetch()
     * @returns Promise resolving to ReedSolomonErasure instance
     *
     * @example
     * ```typescript
     * const rs = await ReedSolomonErasure.fromResponse(
     *   fetch('/path/to/reed_solomon_erasure_bg.wasm')
     * );
     * ```
     */
    static async fromResponse(source) {
        const response = await source;
        // Check if instantiateStreaming is available (not in all environments)
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            try {
                const { instance } = await WebAssembly.instantiateStreaming(response.clone());
                return new ReedSolomonErasure(instance.exports);
            }
            catch {
                // Fall back to arrayBuffer method if streaming fails
            }
        }
        // Fallback: load as ArrayBuffer
        const bytes = await response.arrayBuffer();
        return ReedSolomonErasure.fromBytes(bytes);
    }
    /**
     * For synchronous instantiation from raw bytes.
     * Works in both Node.js and Browser environments.
     *
     * @param bytes - The WASM binary as BufferSource (ArrayBuffer, Uint8Array, etc.)
     * @returns ReedSolomonErasure instance
     *
     * @example
     * ```typescript
     * // Node.js
     * const wasmBytes = fs.readFileSync('./reed_solomon_erasure_bg.wasm');
     * const rs = ReedSolomonErasure.fromBytes(wasmBytes);
     *
     * // Browser (after fetching)
     * const response = await fetch('./reed_solomon_erasure_bg.wasm');
     * const bytes = await response.arrayBuffer();
     * const rs = ReedSolomonErasure.fromBytes(bytes);
     * ```
     */
    static fromBytes(bytes) {
        const module = new WebAssembly.Module(bytes);
        const instance = new WebAssembly.Instance(module);
        return new ReedSolomonErasure(instance.exports);
    }
    /**
     * Synchronous factory for Node.js environment only.
     * Loads WASM from the package's dist directory.
     *
     * @throws Error if called in browser environment
     * @returns ReedSolomonErasure instance
     *
     * @example
     * ```typescript
     * // Node.js only
     * const rs = ReedSolomonErasure.fromCurrentDirectorySync();
     * ```
     */
    static fromCurrentDirectorySync() {
        if (isBrowser) {
            throw new Error('fromCurrentDirectorySync() is not available in browser. ' +
                'Use fromCurrentDirectory() or fromResponse() instead.');
        }
        const wasmBytes = loadWasmBytesSyncNode(`${__dirname}/reed_solomon_erasure_bg.wasm`);
        return ReedSolomonErasure.fromBytes(wasmBytes);
    }
    /**
     * Takes a contiguous array of bytes that contain space for `data_shards + parity_shards`
     * shards with `data_shards` shards containing data and fills additional `parity_shards`
     * with parity information that can be later used to reconstruct data in case of corruption.
     *
     * @param shards - Contiguous byte array with space for all shards
     * @param dataShards - Number of data shards
     * @param parityShards - Number of parity shards to generate
     *
     * @returns One of `RESULT_*` constants; if `RESULT_OK` then parity shards were updated in `shards` in-place
     *
     * @example
     * ```typescript
     * const dataShards = 4;
     * const parityShards = 2;
     * const shardSize = 1024;
     * const totalSize = shardSize * (dataShards + parityShards);
     *
     * const shards = new Uint8Array(totalSize);
     * // Fill first 4 shards with data...
     *
     * const result = rs.encode(shards, dataShards, parityShards);
     * if (result === ReedSolomonErasure.RESULT_OK) {
     *   // Parity shards are now filled in
     * }
     * ```
     */
    encode(shards, dataShards, parityShards) {
        const exports = this.exports;
        const shardsLength = shards.length;
        const shardsPointer = exports.__wbindgen_malloc(shardsLength);
        this.getUint8Memory().set(shards, shardsPointer);
        const shardSize = shardsLength / (dataShards + parityShards);
        const result = exports.encode(shardsPointer, shardsLength, dataShards, parityShards);
        if (result === ReedSolomonErasure.RESULT_OK) {
            shards.set(this.getUint8Memory().subarray(shardsPointer + shardSize * dataShards, shardsPointer + shardsLength), shardSize * dataShards);
        }
        exports.__wbindgen_free(shardsPointer, shardsLength);
        return result;
    }
    /**
     * Takes a contiguous array of bytes that contain `data_shards + parity_shards` shards
     * and tries to reconstruct data shards if they are broken and whenever possible using
     * information from `shardsAvailable`.
     *
     * @param shards - Contiguous byte array containing all shards
     * @param dataShards - Number of data shards
     * @param parityShards - Number of parity shards
     * @param shardsAvailable - Array of booleans indicating which shards are valid (true) or corrupted (false)
     *
     * @returns One of `RESULT_*` constants; if `RESULT_OK` then data shards were reconstructed in `shards` in-place
     *
     * @example
     * ```typescript
     * const dataShards = 4;
     * const parityShards = 2;
     *
     * // Simulate corruption of shards 1 and 2
     * const shardsAvailable = [true, false, false, true, true, true];
     *
     * const result = rs.reconstruct(shards, dataShards, parityShards, shardsAvailable);
     * if (result === ReedSolomonErasure.RESULT_OK) {
     *   // Data shards are now reconstructed
     * }
     * ```
     */
    reconstruct(shards, dataShards, parityShards, shardsAvailable) {
        const exports = this.exports;
        const shardsLength = shards.length;
        const shardsPointer = exports.__wbindgen_malloc(shardsLength);
        this.getUint8Memory().set(shards, shardsPointer);
        const shardsAvailableLength = shardsAvailable.length;
        const shardsAvailablePointer = exports.__wbindgen_malloc(shardsAvailableLength);
        this.getUint8Memory().set(shardsAvailable.map((value) => value ? 1 : 0), shardsAvailablePointer);
        const shardSize = shardsLength / (dataShards + parityShards);
        const result = exports.reconstruct(shardsPointer, shardsLength, dataShards, parityShards, shardsAvailablePointer, shardsAvailableLength);
        if (result === ReedSolomonErasure.RESULT_OK) {
            shards.set(this.getUint8Memory().subarray(shardsPointer, shardsPointer + shardSize * dataShards));
        }
        exports.__wbindgen_free(shardsPointer, shardsLength);
        exports.__wbindgen_free(shardsAvailablePointer, shardsAvailableLength);
        return result;
    }
    /**
     * Get human-readable description for a result code.
     *
     * @param code - Result code from encode() or reconstruct()
     * @returns Human-readable error message
     */
    static getResultMessage(code) {
        const messages = {
            0: 'OK',
            1: 'Too few shards',
            2: 'Too many shards',
            3: 'Too few data shards',
            4: 'Too many data shards',
            5: 'Too few parity shards',
            6: 'Too many parity shards',
            7: 'Too few buffer shards',
            8: 'Too many buffer shards',
            9: 'Incorrect shard size',
            10: 'Too few shards present for reconstruction',
            11: 'Empty shard',
            12: 'Invalid shard flags',
            13: 'Invalid index',
        };
        return messages[code] || `Unknown error code: ${code}`;
    }
    /**
     * Load WASM from a base64-encoded string.
     * Useful for embedding WASM directly in JavaScript bundles.
     *
     * @param base64 - Base64-encoded WASM binary
     */
    static fromBase64(base64) {
        const binaryString = atob(base64);
        const bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return ReedSolomonErasure.fromBytes(bytes);
    }
    /**
     * Get the Uint8Array view of WASM memory, with caching for performance.
     */
    getUint8Memory() {
        let cachegetUint8Memory = this.memoryCache;
        if (cachegetUint8Memory === null ||
            cachegetUint8Memory.buffer !== this.exports.memory.buffer) {
            cachegetUint8Memory = new Uint8Array(this.exports.memory.buffer);
            this.memoryCache = cachegetUint8Memory;
        }
        return cachegetUint8Memory;
    }
}
exports.ReedSolomonErasure = ReedSolomonErasure;
ReedSolomonErasure.RESULT_OK = 0;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_SHARDS = 1;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_SHARDS = 2;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_DATA_SHARDS = 3;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_DATA_SHARDS = 4;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_PARITY_SHARDS = 5;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_PARITY_SHARDS = 6;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_BUFFER_SHARDS = 7;
ReedSolomonErasure.RESULT_ERROR_TOO_MANY_BUFFER_SHARDS = 8;
ReedSolomonErasure.RESULT_ERROR_INCORRECT_SHARD_SIZE = 9;
ReedSolomonErasure.RESULT_ERROR_TOO_FEW_SHARDS_PRESENT = 10;
ReedSolomonErasure.RESULT_ERROR_EMPTY_SHARD = 11;
ReedSolomonErasure.RESULT_ERROR_INVALID_SHARD_FLAGS = 12;
ReedSolomonErasure.RESULT_ERROR_INVALID_INDEX = 13;
// Default export for convenience
exports.default = ReedSolomonErasure;
